packages/data/src/namespace-store.js

```diff
@@ -103,7 +103,24 @@ function createReduxStore( reducer, key, registry ) {
  * @return {Object}           Selectors mapped to the redux store provided.
  */
 function mapSelectors( selectors, store ) {
-	const createStateSelector = ( selector ) => ( ...args ) => selector( store.getState(), ...args );
+	const createStateSelector = ( selector ) => function runSelector() {
+		// This function is an optimized implementation of:
+		//
+		//   selector( store.getState(), ...arguments )
+		//
+		// Where the above would incur an `Array#concat` in its application,
+		// the logic here instead efficiently constructs an arguments array via
+		// direct assignment.
+		const argsLength = arguments.length;
+		const args = new Array( argsLength + 1 );
+		args[ 0 ] = store.getState();
+		for ( let i = 0; i < argsLength; i++ ) {
+			args[ i + 1 ] = arguments[ i ];
+		}
```

Comment:
Maybe this could be extracted to a utility function `concat` for clarity?

Reply:
`arguments` and utility functions don't get along, though from some benchmarks I've found, it's becoming less true over time.

- https://github.com/petkaantonov/bluebird/wiki/Optimization-killers#3-managing-arguments
- https://stackoverflow.com/a/30237178
- https://www.jstips.co/en/javascript/avoid-modifying-or-passing-arguments-into-other-functions-it-kills-optimization/

Reply:
I’m fine about leaving it inline as is.

I was wondering whether updating the return statement would also have an impact: 
```js
return selector.apply( null, args );
```

Maybe Babel does it already, I’m not sure but it might be worth checking.

Reply:
Yeah, the transpiled output has essentially the same result:

https://babeljs.io/repl#?babili=false&browsers=&build=&builtIns=false&spec=false&loose=false&code_lz=MYewdgzgLgBMBOBTAhlRBlKqOIDaOChHhgF4YAKGCPAokgSjID4YAzAVzEIEtwZ4XdLULEKTAN4AoAJChIsZPADmEADKIwyqAAsyMJco4BbTVAgA6fFt0BuWfOgGVEfWEQB3GAEF48ZACeVIbqmtp6ANQwAIwwDPYyIQDaMAAMMAC6-tDEiBbKiFCY2OIJbMSUMPiwPPqptjC1ADzOqho2Og08EVGSsokuKbVRsVnkhiZmEEOZCQC-UrJIUBzwYNQi9FQWOyFx9nO2QA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=module&lineWrap=true&presets=es2015%2Creact%2Cstage-2&prettier=false&targets=&version=6.26.0&envVersion=

Reply:
Andrew, thanks for checking. Babel, good job! 👏